### PR TITLE
1110 oppanel

### PR DESCRIPTION
**Pull request by sagisin**

### CODE REVIEW

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 1]
```diff
@@ -689,6 +689,46 @@ void pldm::responder::oem_ibm_platform::Handler::setSurvTimer(uint8_t tid,
...
+        auto& bus = pldm::utils::DBusHandler::getBus();
+        auto method = bus.new_method_call(service, objPath, dbusInterface,
+                                          dbusMethod);
+        auto val = std::get_if<std::vector<uint8_t>>(&value);
+        method.append(*val);
+        bus.call_noreply(method);
+    }
+    catch (const std::exception& e)
+    {
+        std::cerr << "Failed to call the D-Bus Method"
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:707]: lg2 log

File: oem/ibm/libpldmresponder/oem_ibm_handler.hpp [hunk 2]
```diff
@@ -297,6 +299,26 @@ class Handler : public oem_platform::Handler
...
+    /** @brief D-Bus Method call to call the Panel D-Bus API
+     *
+     * @param[in] service - Service name for the D-Bus method
+     * @param[in] objPath - The D-Bus object path
+     * @param[in] dbusMethod - The Method name to be invoked
+     * @param[in] dbusInterface - The D-Bus interface
+     * @param[in] value - The value to be passed as argument
+     *            to D-Bus method
+     */
+    void setBitmapMethodCall(const char* service, const char* objPath,
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.hpp:318]: can we pass by reference?

File: oem/ibm/libpldmresponder/oem_ibm_handler.hpp [hunk 3]
```diff
@@ -333,6 +355,10 @@ class Handler : public oem_platform::Handler
     /** @brief D-Bus property changed signal match for CurrentPowerState*/
     std::unique_ptr<sdbusplus::bus::match_t> chassisOffMatch;
 
+    // pdr_utils::Repo pdrRepo;
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.hpp:358]: remove

File: host-bmc/host_pdr_handler.cpp [hunk 4]
```diff
@@ -652,12 +652,62 @@ void HostPDRHandler::processHostPDRs(mctp_eid_t /*eid*/,
...
-                        throw std::runtime_error("Failed to add PDR");
+                        isHostPdrModified = false;
+
+                        pldm_delete_by_record_handle(repo, rh, true);
+
+                        pldm_pdr_add_check(repo, pdr.data(), respCount, true,
+                                           pdrTerminusHandle, &rh);
+
+                        if ((pdrHdr->type == PLDM_STATE_EFFECTER_PDR) &&
+                            (oemPlatformHandler != nullptr))
```
Comment (Pavithrab7) [inline host-bmc/host_pdr_handler.cpp:665]: (oemPlatformHandler)

File: host-bmc/host_pdr_handler.cpp [hunk 5]
```diff
@@ -652,12 +652,62 @@ void HostPDRHandler::processHostPDRs(mctp_eid_t /*eid*/,
...
+                    }
+                    // We need to look for an optimal solution for this, we are
+                    // unexpectedly entering this path when we receive multiple
+                    // modified PDR repo change events
+                    else if ((isHostPdrModified != true) &&
+                             (modifiedCounter == 0))
+                    {
+                        pldm_delete_by_record_handle(repo, rh, true);
+
+                        pldm_pdr_add_check(repo, pdr.data(), respCount, true,
```
Comment (Pavithrab7) [inline host-bmc/host_pdr_handler.cpp:700]: do we want to handle rc here as well?

File: host-bmc/host_pdr_handler.cpp [hunk 6]
```diff
@@ -652,12 +652,62 @@ void HostPDRHandler::processHostPDRs(mctp_eid_t /*eid*/,
...
-                    if (rc)
+                    if ((isHostPdrModified == true) || !(modifiedCounter == 0))
                     {
-                        // pldm_pdr_add() assert()ed on failure to add a PDR.
-                        throw std::runtime_error("Failed to add PDR");
+                        isHostPdrModified = false;
+
+                        pldm_delete_by_record_handle(repo, rh, true);
+
+                        pldm_pdr_add_check(repo, pdr.data(), respCount, true,
```
Comment (Pavithrab7) [inline host-bmc/host_pdr_handler.cpp:661]: do we want to handle rc from `pldm_pdr_add_check` here as well?

**Review by Pavithrab7 — CHANGES_REQUESTED** [review]

**Review by manojkiraneda — CHANGES_REQUESTED** [review]
Agree with all the comments that @Pavithrab7 gave.

**Comment by Pavithrab7** [comment]
I do not see decrementing the `modifiedCounter` value anywhere in the code? we must do that right?

File: host-bmc/host_pdr_handler.cpp (hunk 6 above)
Comment (sagisin) [reply host-bmc/host_pdr_handler.cpp:661]: done

**Comment by sagisin** [comment]
> I do not see decrementing the `modifiedCounter` value anywhere in the code? we must do that right?

its done

File: host-bmc/host_pdr_handler.cpp [hunk 7]
```diff
@@ -664,12 +664,71 @@ void HostPDRHandler::processHostPDRs(mctp_eid_t /*eid*/,
...
+                    else if ((isHostPdrModified != true) &&
+                             (modifiedCounter == 0))
+                    {
+                        pldm_delete_by_record_handle(repo, rh, true);
+
+                        rc = pldm_pdr_add_check(repo, pdr.data(), respCount,
+                                                true, pdrTerminusHandle, &rh);
+                        if (rc)
+                        {
+                            throw std::runtime_error("Failed to add PDR");
```
Comment (Pavithrab7) [inline host-bmc/host_pdr_handler.cpp:721]: can we improve this trace? `Failed to add PDR when isHostPdrModified is not true` something like this?
As even below we have same trace.

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 8]
```diff
@@ -1066,6 +1066,49 @@ void pldm::responder::oem_ibm_platform::Handler::processSAIUpdate()
...
+                                          dbusInterface.c_str(),
+                                          dbusMethod.c_str());
+        auto val = std::get_if<std::vector<uint8_t>>(&value);
+        method.append(*val);
+        bus.call_noreply(method);
+    }
+    catch (const std::exception& e)
+    {
+        error("Failed to call the D-Bus Method ERROR={ERR_EXCEP}", "ERR_EXCEP",
+              e.what());
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:1088]: just a `e`

**Review by Pavithrab7 — CHANGES_REQUESTED** [review]

File: host-bmc/host_pdr_handler.cpp (hunk 7 above)
Comment (sagisin) [reply host-bmc/host_pdr_handler.cpp]: done

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp (hunk 8 above)
Comment (sagisin) [reply oem/ibm/libpldmresponder/oem_ibm_handler.cpp]: done

**Review by manojkiraneda — APPROVED** [review]

File: host-bmc/host_pdr_handler.cpp [hunk 9]
```diff
@@ -664,12 +664,73 @@ void HostPDRHandler::processHostPDRs(mctp_eid_t /*eid*/,
...
+                        isHostPdrModified = false;
+
+                        pldm_delete_by_record_handle(repo, rh, true);
+
+                        rc = pldm_pdr_add_check(repo, pdr.data(), respCount,
+                                                true, pdrTerminusHandle, &rh);
+                        if (rc)
+                        {
+                            throw std::runtime_error(
+                                "Failed to add PDR when isHostPdrModified is not true");
```
Comment (Pavithrab7) [inline host-bmc/host_pdr_handler.cpp:678]: this is when its true right?

**Review by Pavithrab7 — CHANGES_REQUESTED** [review]

**Review by Pavithrab7 — APPROVED** [review]

File: host-bmc/host_pdr_handler.cpp (hunk 9 above)
Comment (sagisin) [reply host-bmc/host_pdr_handler.cpp]: done

**Pull request merged**